src/lib/Parser.hs

```diff
@@ -341,17 +357,21 @@ interfaceInstance = do
   (p, pos) <- withPos letPat
   ann <- annot uType
   keyWord WhereKW
-  rhs <- mkRhs ann <$> (withSrc $ URecord <$> interfaceRecordFields "=")
+  rhs <- mkConstructorCall ann <$> (withSrc $ (URecord . NoExt) <$>
+      interfaceRecordFields "=")
   let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
   return $ ULet InstanceLet (p, ann') rhs'
   where
     -- Here, we are traversing the type annotation to retrieve the name of
     -- the interface and generate its corresponding constructor.
-    mkRhs (WithSrc _ (UPi _ _ lrhs))      rhs = mkRhs lrhs rhs
-    mkRhs (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) rhs =
-      (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` rhs
-    -- We let this case through to fail during type checking.
-    mkRhs _ rhs = rhs
+    mkConstructorCall (WithSrc _ (UPi _ ClassArrow typ)) record =
+      mkConstructorCall typ record
+    mkConstructorCall (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) record =
+      (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` record
+    mkConstructorCall (WithSrc _ (UApp _ func _)) record =
+      mkConstructorCall func record
+    -- We let this case through to fail during typechecking
+    mkConstructorCall _ record = record
     var s = WithSrc Nothing $ UVar $ mkName s :> ()
 
 oneLabeledItem
```

Comment:
Weird! I distinctly remember writing a comment about this function on the last PR but it seems to have vanished.

My opinion is that the functionality here isn't common enough for it to make sense as a standalone helper function, since the syntax of a record or variant is quite different than the syntax of parsing an interface. I would prefer to have this inlined in the two use sites (i.e. reverted to how it was before #249, and then just inlining the two or three common lines into `interfaceRecordFields`). Most of the complexity here has to do with handling variable punning, which I don't think you should be doing anyway.